File: use-cases/quality-control/test-pattern.py
"""
Simple Quality Control Test
Test the pattern with a sample product image
"""

import os
import sys
import json
from pathlib import Path

sys.path.append(str(Path(__file__).parent.parent.parent / "examples" / "common"))
from imggo_client import ImgGoClient

def main():
    print("="*60)
    print("TESTING QUALITY CONTROL PATTERN")
    print("="*60)

    if not os.getenv("IMGGO_API_KEY"):
        print("\nX Error: IMGGO_API_KEY not set")
        sys.exit(1)

    # Check pattern ID - from env var or pattern_id.txt
    pattern_id = os.getenv("QUALITY_CONTROL_PATTERN_ID")
    if not pattern_id:
        pattern_file = Path(__file__).parent / "pattern_id.txt"
        if pattern_file.exists():
            pattern_id = pattern_file.read_text().strip()
    if not pattern_id:
        print("\nX Error: QUALITY_CONTROL_PATTERN_ID not set")
        print("Run create-pattern.py first to create a pattern")
        sys.exit(1)

    # Use inventory image for quality inspection
    test_image = Path(__file__).parent.parent.parent / "examples" / "test-images" / "inventory1.jpg"
    if not test_image.exists():
        print(f"\nX Error: Test image not found: {test_image}")
        sys.exit(1)

    print(f"\nPattern ID: {pattern_id}")
    print(f"Test Image: {test_image.name}\n")

    try:
        client = ImgGoClient()
        print("Processing quality inspection image...")

        result = client.process_image(
            image_path=str(test_image),
            pattern_id=pattern_id
        )

        print("V Processing completed!\n")

        outputs_dir = Path(__file__).parent / "outputs"
        outputs_dir.mkdir(exist_ok=True)

        # JSON format output
        output_file = outputs_dir / "quality_control_output.json"
        try:
            qc_data = result if isinstance(result, dict) else json.loads(result)
            with open(output_file, 'w', encoding='utf-8') as f:
                json.dump(qc_data, f, indent=2)

            print(f"V Output saved to: {output_file}\n")
            print("="*60)
            print("EXTRACTED QC DATA")
            print("="*60)
            print(json.dumps(qc_data, indent=2))

        except json.JSONDecodeError:
            # Not JSON, save as text
            with open(output_file, 'w', encoding='utf-8') as f:
                f.write(result if isinstance(result, str) else str(result))
            print(f"V Output saved to: {output_file}\n")
            print("="*60)
            print("EXTRACTED DATA")
            print("="*60)
            print(result)

        print("\nV Test completed successfully!")

    except Exception as e:
        print(f"\nX Error: {e}")
        import traceback
        traceback.print_exc()
        sys.exit(1)

if __name__ == "__main__":
    main()

File: use-cases/quality-control/test-pattern.ts
/**
 * Quality Control Test - TypeScript/Node.js Example
 * Demonstrates ImgGo API usage with TypeScript
 */

import * as fs from 'fs';
import * as path from 'path';
import * as https from 'https';

const IMGGO_API_KEY = process.env.IMGGO_API_KEY;
const BASE_URL = 'img-go.com';

interface JobResponse {
    success: boolean;
    data: {
        job_id?: string;
        status?: string;
        manifest?: Record<string, unknown>;
        result?: Record<string, unknown>;
    };
}

function getPatternId(): string | null {
    const envId = process.env.QUALITY_CONTROL_PATTERN_ID;
    if (envId) return envId;

    const patternFile = path.join(__dirname, 'pattern_id.txt');
    if (fs.existsSync(patternFile)) {
        return fs.readFileSync(patternFile, 'utf-8').trim();
    }
    return null;
}

async function submitImage(patternId: string, imagePath: string): Promise<string> {
    return new Promise((resolve, reject) => {
        const imageData = fs.readFileSync(imagePath);
        const boundary = '----FormBoundary' + Math.random().toString(36).substring(2);

        const body = Buffer.concat([
            Buffer.from(`--${boundary}\r\n`),
            Buffer.from(`Content-Disposition: form-data; name="image"; filename="${path.basename(imagePath)}"\r\n`),
            Buffer.from('Content-Type: image/jpeg\r\n\r\n'),
            imageData,
            Buffer.from(`\r\n--${boundary}--\r\n`)
        ]);

        const options: https.RequestOptions = {
            hostname: BASE_URL,
            path: `/api/patterns/${patternId}/ingest`,
            method: 'POST',
            headers: {
                'Authorization': `Bearer ${IMGGO_API_KEY}`,
                'Content-Type': `multipart/form-data; boundary=${boundary}`,
                'Content-Length': body.length,
                'Idempotency-Key': `qc-ts-${Date.now()}`
            }
        };

        const req = https.request(options, (res) => {
            let data = '';
            res.on('data', chunk => data += chunk);
            res.on('end', () => {
                try {
                    const response: JobResponse = JSON.parse(data);
                    if (response.data?.job_id) {
                        resolve(response.data.job_id);
                    } else {
                        reject(new Error(`No job_id in response: ${data}`));
                    }
                } catch (e) {
                    reject(new Error(`Failed to parse response: ${data}`));
                }
            });
        });

        req.on('error', reject);
        req.write(body);
        req.end();
    });
}

async function pollForResult(jobId: string, maxAttempts = 30): Promise<Record<string, unknown>> {
    for (let i = 0; i < maxAttempts; i++) {
        const result = await new Promise<JobResponse>((resolve, reject) => {
            const options: https.RequestOptions = {
                hostname: BASE_URL,
                path: `/api/jobs/${jobId}`,
                method: 'GET',
                headers: {
                    'Authorization': `Bearer ${IMGGO_API_KEY}`
                }
            };

            const req = https.request(options, (res) => {
                let data = '';
                res.on('data', chunk => data += chunk);
                res.on('end', () => {
                    try {
                        resolve(JSON.parse(data));
                    } catch (e) {
                        reject(new Error(`Failed to parse response: ${data}`));
                    }
                });
            });

            req.on('error', reject);
            req.end();
        });

        const status = result.data?.status;
        if (status === 'succeeded' || status === 'completed') {
            return result.data?.manifest || result.data?.result || result.data;
        } else if (status === 'failed') {
            throw new Error('Job failed');
        }

        console.log(`Status: ${status}, waiting... (${i + 1}/${maxAttempts})`);
        await new Promise(resolve => setTimeout(resolve, 2000));
    }
    throw new Error('Timeout waiting for result');
}

async function main(): Promise<void> {
    console.log('='.repeat(60));
    console.log('TESTING QUALITY CONTROL PATTERN (TypeScript)');
    console.log('='.repeat(60));

    if (!IMGGO_API_KEY) {
        console.log('\nX Error: IMGGO_API_KEY not set');
        process.exit(1);
    }

    const patternId = getPatternId();
    if (!patternId) {
        console.log('\nX Error: QUALITY_CONTROL_PATTERN_ID not set');
        console.log('Run create-pattern.py first to create a pattern');
        process.exit(1);
    }

    const testImage = path.join(__dirname, '../../examples/test-images/inventory1.jpg');
    if (!fs.existsSync(testImage)) {
        console.log(`\nX Error: Test image not found: ${testImage}`);
        process.exit(1);
    }

    console.log(`\nPattern ID: ${patternId}`);
    console.log(`Test Image: inventory1.jpg\n`);

    try {
        console.log('Processing quality inspection image...');
        const jobId = await submitImage(patternId, testImage);
        console.log(`Job submitted: ${jobId}`);

        const result = await pollForResult(jobId);
        console.log('V Processing completed!\n');

        // Save output
        const outputDir = path.join(__dirname, 'outputs');
        if (!fs.existsSync(outputDir)) {
            fs.mkdirSync(outputDir);
        }

        const outputFile = path.join(outputDir, 'quality_control_ts_output.json');
        fs.writeFileSync(outputFile, JSON.stringify(result, null, 2));

        console.log(`V Output saved to: ${outputFile}\n`);
        console.log('='.repeat(60));
        console.log('QC INSPECTION RESULTS');
        console.log('='.repeat(60));
        console.log(JSON.stringify(result, null, 2));
        console.log('\nV Test completed successfully!');

    } catch (error) {
        console.log(`\nX Error: ${error}`);
        process.exit(1);
    }
}

main();
